src/opam.ml

```diff
@@ -127,14 +129,30 @@ end = struct
 
   let of_switch opam switch =
     let open Promise.Syntax in
-    let path = Switch.path opam switch in
-    let switch_state_filepath = Opam_path.switch_state path |> Path.to_string in
-    let+ file_content = Fs.readFile switch_state_filepath in
-    let { OpamParserTypes.file_contents; _ } =
-      OpamParser.FullPos.string file_content switch_state_filepath
-      |> OpamParser.FullPos.to_opamfile
+    let switch_state_filepath =
+      Switch.path opam switch |> Opam_path.switch_state |> Path.to_string
     in
-    file_contents
+    let+ file_res =
+      Promise.Result.from_catch (Fs.readFile switch_state_filepath)
```

Comment:
I'm surprised we don't need a thunk to wrap the read operation here. Can't `Fs.readFile` raise immediately as well?

Reply:
To my best knowledge, raising an exception would violate nodejs-way of working with async code. I think the user of this function can safely expect the error in `err` argument in the callback: 

```
export function readFile(
    path: PathLike | number,
    options: { encoding: string; flag?: string | undefined } | string,
    callback: (err: NodeJS.ErrnoException | null, data: string) => void
): void;
```

When it comes to `Promise.Result.from_catch` and its behavior, I think it does what it should since `(<function call returning promise>).catch(...)` doesn't catch the exception if the function call throws one outside of promise, e.g.,

```js
function delayedPrint() { 
  throw new Error("this is thrown and not caught by `.catch`")
  return new Promise(resolve => 
         setTimeout(() => { console.log("not printed"); resolve() }, 100));
}

delayedPrint().catch(e => console.log("caught e: " + e))
```